### PR TITLE
Bluetooth: Controller: Implement Periodic Advertiser List

### LINKED ISSUE #38316: Synchronize multiple DF TX devices in the DF Connectionless RX Example "Periodic Advertising list"
I got an Unknown Command error while working with the periodic advertising list feature #29127

When I tried to add the device to the Periodic Advertising list or clear the Periodic Advertising list, **error -5** shows over the terminal after deep looking at the source code found the reason for such error is that commands (BT_HCI_OP_LE_ADD_DEV_TO_PER_ADV_LIST, BT_HCI_OP_LE_REM_DEV_FROM_PER_ADV_LIST, BT_HCI_OP_LE_CLEAR_PER_ADV_LIST, BT_HCI_OP_LE_READ_PER_ADV_LIST_SIZE) are not defined/handled in the **static int controller_cmd_handle(uint16_t  ocf, struct net_buf *cmd, struct net_buf **evt, void **node_rx)** function.

Is there any way to solve such a bug, kindly guide me?

### PR DESCRIPTION
**Pull request by cvinayak**
Implement support for Periodic Advertiser List to be used
in LE Periodic Advertising Create Sync command.

Fixes #38316.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

### CODE REVIEW

File: subsys/bluetooth/controller/ll_sw/ull_filter.c [hunk 1]
```diff
@@ -138,10 +140,29 @@ static uint32_t filter_remove(struct lll_filter *filter, uint8_t addr_type,
...
-static void filter_insert(struct lll_filter *filter, int index, uint8_t addr_type,
-			   uint8_t *bdaddr);
+static void filter_insert(struct lll_filter *filter, int index,
+			  uint8_t addr_type, uint8_t *bdaddr);
 static void filter_clear(struct lll_filter *filter);
 
+#if defined(CONFIG_BT_CTLR_SYNC_PERIODIC)
+#define PL_ADDR_MATCH(type, addr) \
+		(pl[i].taken && \
+		(pl[i].id_addr_type == (type & 0x1)) && \
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/ull_filter.c:150]: Do we have a #define for the type values, instead of 0x1?

File: subsys/bluetooth/controller/ll_sw/ull_filter.c [hunk 2]
```diff
@@ -470,7 +513,138 @@ uint8_t ll_priv_mode_set(bt_addr_le_t *id_addr, uint8_t mode)
...
+{
+	/* FIXME: Check and fail if Periodic Advertising Create Sync is pending.
+	 */
+
+	pl_clear();
+
+	return 0;
+}
+
+uint8_t ll_pl_add(bt_addr_le_t *addr, uint8_t sid)
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/ull_filter.c:534]: For this and the following functions: 

```suggestion
uint8_t ll_pl_add(const bt_addr_le_t *addr, uint8_t sid)
```

File: subsys/bluetooth/controller/ll_sw/ull_scan_aux.c [hunk 3]
```diff
@@ -270,9 +271,52 @@ void ull_scan_aux_setup(memq_link_t *link, struct node_rx_hdr *rx)
 
 	if (h->adv_addr) {
 #if defined(CONFIG_BT_CTLR_SYNC_PERIODIC)
-		if (sync && (pdu->tx_addr == scan->per_scan.adv_addr_type) &&
-		    !memcmp(ptr, scan->per_scan.adv_addr, BDADDR_SIZE)) {
-			scan->per_scan.state = LL_SYNC_STATE_ADDR_MATCH;
+		/* Check if Periodic Advertising Synchronization to be created
+		 */
+		if (sync) {
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/ull_scan_aux.c:276]: This adds another 40 lines to a function that's almost 500 lines long. I strongly suggest that we look into splitting this function.

**Review by Thalley — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/controller/Kconfig [hunk 4]
```diff
@@ -384,14 +384,14 @@ config BT_CTLR_PRIVACY
 	  Enable support for Bluetooth v4.2 LE Controller-based Privacy feature
 	  in the Controller.
 
-config BT_CTLR_WL_SIZE
-	int "LE Controller-based Privacy White List size"
+config BT_CTLR_AL_SIZE
```
[the remark this answers is not in the document]
Comment (cvinayak) [reply subsys/bluetooth/controller/Kconfig:387]: Changed to ACCEPT_LIST size, as this is common for device address match and private address resolution feature.

File: subsys/bluetooth/controller/Kconfig [hunk 5]
```diff
@@ -497,6 +497,15 @@ config BT_CTLR_SYNC_PERIODIC
 config BT_CTLR_SYNC_PERIODIC
 	bool "LE Periodic Advertising in Synchronization State [EXPERIMENTAL]" if BT_LL_SW_SPLIT
 
+config BT_CTLR_PL_SIZE
```
[the remark this answers is not in the document]
Comment (cvinayak) [reply subsys/bluetooth/controller/Kconfig:500]: Changed to SYNC_PERIODIC_ADV_LIST_SIZE

File: subsys/bluetooth/controller/Kconfig [hunk 6]
```diff
@@ -497,6 +497,15 @@ config BT_CTLR_SYNC_PERIODIC
...
+	int "LE Periodic Advertiser List size"
+	depends on BT_CTLR_FILTER && BT_CTLR_SYNC_PERIODIC
```
[the remark this answers is not in the document]
Comment (cvinayak) [reply subsys/bluetooth/controller/Kconfig:502]: Have made periodic advertiser list independent of filter accept list and privacy features.

Users can select:
1. Only filter accept list, or
2. Only periodic advertiser list, or
3. Filter accept list and periodic advertiser list, or
4. Filter accept list and privacy, or
5. Filter accept list and periodic advertiser list and privacy

The option, Periodic advertiser list and privacy without Filter accept list is not covered, but can be done in the future (too much conditional compile combinations already)

File: subsys/bluetooth/controller/ll_sw/ull_filter.c (hunk 1 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_filter.c:150]: `0x01` is used as 1-bit mask, as `id_addr_type` is 1-bit size.

File: subsys/bluetooth/controller/ll_sw/ull_scan_aux.c (hunk 3 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_scan_aux.c:276]: I will refactor in a separate commit in this PR

File: subsys/bluetooth/controller/ll_sw/lll_filter.h [hunk 7]
```diff
@@ -4,65 +4,89 @@
...
 #else
-#error WL_SIZE must be <= 16
+#error AL_SIZE must be <= 16
 #endif
-	uint8_t  bdaddr[WL_SIZE][BDADDR_SIZE];
+	uint8_t  bdaddr[AL_SIZE][BDADDR_SIZE];
+};
+
+/* Accept list */
+struct lll_acceptlist {
```
[the remark this answers is not in the document]
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/lll_filter.h]: changed

File: subsys/bluetooth/controller/ll_sw/lll_filter.h [hunk 8]
```diff
@@ -4,65 +4,89 @@
...
+	uint8_t   rl_idx;
 };
 
-/* Whitelist peer list */
-struct lll_whitelist {
-	uint8_t      taken:1;
-	uint8_t      id_addr_type:1;
-	uint8_t      rl_idx;
+/* Periodic Advertising List */
+struct lll_pa_acceptlist {
```
[the remark this answers is not in the document]
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/lll_filter.h]: changed

File: subsys/bluetooth/controller/ll_sw/ull_filter.c (hunk 2 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_filter.c]: done

File: subsys/bluetooth/controller/ll_sw/ull_scan_aux.c (hunk 3 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_scan_aux.c:275]: done

File: subsys/bluetooth/controller/ll_sw/ull_filter.c [hunk 9]
```diff
@@ -176,17 +199,15 @@ uint8_t ll_wl_clear(void)
...
+	filter_clear(&hw_al);
 #endif /* CONFIG_BT_CTLR_PRIVACY */
 
-	wl_anon = 0U;
-
 	return 0;
 }
 
-uint8_t ll_wl_add(bt_addr_le_t *addr)
+uint8_t ll_al_add(bt_addr_le_t *addr)
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/ull_filter.c:210]: Possibly outside the scope of this PR, but I suppose that we can apply `const` to some of the accept list functions as well.

File: subsys/bluetooth/controller/ll_sw/ull_filter.c [hunk 10]
```diff
@@ -470,7 +515,140 @@ uint8_t ll_priv_mode_set(bt_addr_le_t *id_addr, uint8_t mode)
...
+	 */
+
+	if (addr->type == ADDR_TYPE_ANON) {
+		return 0;
+	}
+
+	return pl_remove(addr, sid);
+}
+
+bool ull_filter_ull_pl_addr_match(uint8_t addr_type, uint8_t *addr)
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/ull_filter.c:561]:
```suggestion
bool ull_filter_ull_pl_addr_match(uint8_t addr_type, const uint8_t *addr)
```

File: subsys/bluetooth/controller/ll_sw/ull_filter.c [hunk 11]
```diff
@@ -470,7 +515,140 @@ uint8_t ll_priv_mode_set(bt_addr_le_t *id_addr, uint8_t mode)
...
+	for (int i = 0; i < PL_SIZE; i++) {
+		if (PL_ADDR_MATCH(addr_type, addr)) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+bool ull_filter_ull_pl_match(uint8_t addr_type, uint8_t *addr, uint8_t sid)
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/ull_filter.c:572]:
```suggestion
bool ull_filter_ull_pl_match(uint8_t addr_type, const uint8_t *addr, uint8_t sid)
```

File: subsys/bluetooth/controller/ll_sw/ull_filter.c [hunk 12]
```diff
@@ -1080,19 +1219,129 @@ static uint32_t filter_remove(struct lll_filter *filter, uint8_t addr_type,
...
+			rl[j].pl = 0U;
+		}
+#endif /* CONFIG_BT_CTLR_PRIVACY */
+
+		pl[i].taken = 0U;
+	}
+}
+
+#if defined(CONFIG_BT_CTLR_PRIVACY)
+static uint8_t pl_addr_find(uint8_t addr_type, uint8_t *addr)
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/ull_filter.c:1255]:
```suggestion
static uint8_t pl_addr_find(uint8_t addr_type, const uint8_t *addr)
```

File: subsys/bluetooth/controller/ll_sw/ull_filter.c [hunk 13]
```diff
@@ -1080,19 +1219,129 @@ static uint32_t filter_remove(struct lll_filter *filter, uint8_t addr_type,
...
+			return i;
+		}
+	}
+
+	return FILTER_IDX_NONE;
+}
+#endif /* CONFIG_BT_CTLR_PRIVACY */
+
+static uint8_t pl_find(uint8_t addr_type, const uint8_t *addr, uint8_t sid,
+		       uint8_t *free_idx)
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/ull_filter.c:1268]: In `ull_filter_rl_find` the `free_idx` argument's type is `uint8_t *const free_idx`. Can/should that be applied here as well?

File: subsys/bluetooth/controller/ll_sw/ull_sync.c [hunk 14]
```diff
@@ -358,6 +370,65 @@ void ull_sync_release(struct ll_sync_set *sync)
...
+
+bool ull_sync_setup_sid_match(struct ll_scan_set *scan, uint8_t sid)
+{
+	return (scan->per_scan.state == LL_SYNC_STATE_ADDR_MATCH) &&
+		(0 ||
+#if defined(CONFIG_BT_CTLR_SYNC_PERIODIC_ADV_LIST)
+		 (scan->per_scan.filter_policy &&
+		  ull_filter_ull_pl_match(scan->per_scan.adv_addr_type,
+					  scan->per_scan.adv_addr, sid)) ||
+#endif /* CONFIG_BT_CTLR_SYNC_PERIODIC_ADV_LIST */
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/ull_sync.c:427]: Can't we do 
```suggestion
		(IS_ENABLED(CONFIG_BT_CTLR_SYNC_PERIODIC_ADV_LIST)
		 (scan->per_scan.filter_policy &&
		  ull_filter_ull_pl_match(scan->per_scan.adv_addr_type,
					  scan->per_scan.adv_addr, sid)) ||
```
here instead?

**Review by Thalley — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/controller/Kconfig [hunk 15]
```diff
@@ -497,6 +497,22 @@ config BT_CTLR_SYNC_PERIODIC
 config BT_CTLR_SYNC_PERIODIC
 	bool "LE Periodic Advertising in Synchronization State [EXPERIMENTAL]" if BT_LL_SW_SPLIT
 
+config BT_CTLR_SYNC_PERIODIC_ADV_LIST
```
Comment (saleh-unikie) [inline subsys/bluetooth/controller/Kconfig:500]: Isn't this config a redundant to the "CONFIG_BT_PER_ADV_SYNC_MAX"?

File: subsys/bluetooth/controller/Kconfig [hunk 16]
```diff
@@ -497,6 +497,22 @@ config BT_CTLR_SYNC_PERIODIC
...
+	bool "LE Periodic Advertiser List support"
+	depends on BT_CTLR_SYNC_PERIODIC
+	default y
+	help
+	  Enable support for LE Periodic Advertiser List support.
+
+config BT_CTLR_SYNC_PERIODIC_ADV_LIST_SIZE
```
Comment (saleh-unikie) [inline subsys/bluetooth/controller/Kconfig:507]: Isn't this config a redundant to the "CONFIG_BT_PER_ADV_SYNC_MAX"?

File: subsys/bluetooth/controller/Kconfig (hunk 15 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/Kconfig:500]: No, that is an API parameter, and this is a Controller feature.

File: subsys/bluetooth/controller/Kconfig (hunk 16 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/Kconfig:507]: No, maximum supported simultaneous synchronizations can be lower than listed number of advertisers; only one advertiser is synchronized for every LE Periodic Advertising Create Sync Command, and _Memory Capacity Exceeded_ error is used to convey insufficient resources to handle any more periodic advertising trains.

Comment (saleh-unikie) [reply subsys/bluetooth/controller/Kconfig:507]: Got it! so setting the "CONFIG_BT_PER_ADV_SYNC_MAX" to a number greater than the "BT_CTLR_SYNC_PERIODIC_ADV_LIST_SIZE" is useless, right?

**Comment by cvinayak** [comment]
@carlescufi Request your attention to this commit  https://github.com/zephyrproject-rtos/zephyr/pull/38338/commits/4bc5d45b388c762b786dc8b999eea4e0e06f8079 in this PR which partly handles inclusive naming related to filter implementation in the Controller. Your review is welcome here in this PR until the commit is ready to be cherry-picked out of this PR (so that other commits in this PR will have reduced rework effort)

File: subsys/bluetooth/controller/Kconfig (hunk 16 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/Kconfig:507]: Maximum simultaneous synchronizations `CONFIG_BT_PER_ADV_SYNC_MAX` can be greater than, says number of trusted peer device stored in a list `CONFIG_BT_CTLR_SYNC_PERIODIC_ADV_LIST_SIZE`. List is just a convenience, to be able to synchronize to any of the listed peers on a first-seen basis. Applications are permitted to use explicit peer device address to create additional synchronizations beyond the number of entries that be accommodated in the Periodic Advertiser List.

File: subsys/bluetooth/controller/ll_sw/ull_filter.c (hunk 9 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_filter.c]: done

File: subsys/bluetooth/controller/ll_sw/ull_filter.c (hunk 10 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_filter.c]: done

File: subsys/bluetooth/controller/ll_sw/ull_filter.c (hunk 11 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_filter.c]: done

File: subsys/bluetooth/controller/ll_sw/ull_filter.c (hunk 12 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_filter.c]: done

File: subsys/bluetooth/controller/ll_sw/ull_filter.c (hunk 13 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_filter.c]: done

File: subsys/bluetooth/controller/ll_sw/ull_sync.c (hunk 14 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_sync.c]: done

File: subsys/bluetooth/controller/ll_sw/lll_filter.h [hunk 17]
```diff
@@ -4,65 +4,89 @@
...
 bool ull_filter_lll_lrpa_used(uint8_t rl_idx);
 extern bt_addr_t *ull_filter_lll_lrpa_get(uint8_t rl_idx);
 extern uint8_t *ull_filter_lll_irks_get(uint8_t *count);
-extern uint8_t ull_filter_lll_rl_idx(bool whitelist, uint8_t devmatch_id);
+extern uint8_t ull_filter_lll_rl_idx(bool is_accept_list, uint8_t devmatch_id);
 extern uint8_t ull_filter_lll_rl_irk_idx(uint8_t irkmatch_id);
-extern bool ull_filter_lll_irk_whitelisted(uint8_t rl_idx);
-extern struct lll_filter *ull_filter_lll_get(bool whitelist);
-extern struct lll_whitelist *ull_filter_lll_whitelist_get(void);
+extern bool ull_filter_lll_irk_accept_listed(uint8_t rl_idx);
```
[the remark this answers is not in the document]
Comment (carlescufi) [reply subsys/bluetooth/controller/ll_sw/lll_filter.h:82]: I changed this in my local commit to `ull_filter_lll_irk_in_fal()`, among many other changes.

File: subsys/bluetooth/controller/Kconfig [hunk 18]
```diff
@@ -391,7 +391,7 @@ config BT_CTLR_FAL_SIZE
 	range 1 8  if  (SOC_COMPATIBLE_NRF || SOC_OPENISA_RV32M1_RISCV32)
 	range 1 16 if !(SOC_COMPATIBLE_NRF || SOC_OPENISA_RV32M1_RISCV32)
 	help
-	  Set the size of the White List for LE Controller-based Privacy.
```
Comment (cvinayak) [inline subsys/bluetooth/controller/Kconfig:394]: @carlescufi Do you want to send a separate PR for this being missed upstream?

File: subsys/bluetooth/controller/hci/hci.c [hunk 19]
```diff
@@ -3341,6 +3344,63 @@ static void le_per_adv_recv_enable(struct net_buf *buf, struct net_buf **evt)
 	ccst = hci_cmd_complete(evt, sizeof(*ccst));
 	ccst->status = status;
 }
+
+#if defined(CONFIG_BT_CTLR_SYNC_PERIODIC_ADV_LIST)
+static void le_add_dev_to_pl(struct net_buf *buf, struct net_buf **evt)
```
Comment (Thalley) [inline subsys/bluetooth/controller/hci/hci.c:3349]: I think I already mentioned this, but after the whitelist was rename filter accept list, and abbreviated to `fal`, I suggest that we use the abbraviation/acronym `pal` for the **p**eriodic **a**dvertiser **l**ist.

Comment (Thalley) [reply subsys/bluetooth/controller/hci/hci.c:3349]: I see that you added another commit in this PR adding `const` to a bunch of functions (which arguably should have been a PR on its own), but to follow that trend, could/should `struct net_buf **evt` have been `struct net_buf **const evt`here as well then?

File: subsys/bluetooth/controller/include/ll.h [hunk 20]
```diff
@@ -9,32 +9,32 @@
...
 
-int ll_init(struct k_sem *sem_rx);
+int ll_init(struct k_sem *const sem_rx);
 void ll_reset(void);
 
 uint8_t ll_set_host_feature(uint8_t bit_number, uint8_t bit_value);
 uint64_t ll_feat_get(void);
 
-uint8_t *ll_addr_get(uint8_t addr_type, uint8_t *p_bdaddr);
+uint8_t *ll_addr_get(uint8_t addr_type, uint8_t *const p_bdaddr);
```
Comment (Thalley) [inline subsys/bluetooth/controller/include/ll.h:18]: iirc, `const <type> *ptr` means that whatever `ptr` points to cannot but changed, but the `ptr` value inside the function may still be changed, which doesn't really affect the caller. 

The `<type> *const ptr` means that the value that is being pointed to can be changed, but the `ptr` value cannot be overriden inside the function, but as this doesn't affect the caller in any way, why did you add this to this and many more functions? My assumption would be such that we don't accidentally change the pointer value in function, but is that it?

File: subsys/bluetooth/controller/ll_sw/lll_filter.h [hunk 21]
```diff
@@ -28,53 +28,73 @@ struct lll_filter {
...
+	uint8_t   lirk:1;
+	uint8_t   dev:1;
+	uint8_t   fal:1;
+
+	uint8_t   id_addr_type:1;
 	bt_addr_t id_addr;
 
-	uint8_t      local_irk[16];
-	uint8_t      pirk_idx;
+	uint8_t   local_irk[16];
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/lll_filter.h:61]: I think I saw a IRK_SIZE #define somewhere in a C file that possibly could be moved to a header file and used here as well.

**Comment by Thalley** [comment]
@cvinayak I feel like this PR is becoming a catch-all for several issue (updates to filter accept list and addition of `const`), and that kind of distracts from what the PR should be doing. 

Since I've already reviewed it, it's okay for me to keep as is, but probably beneficial for others if those changes were split to 3 PRs.

File: subsys/bluetooth/controller/include/ll.h (hunk 20 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/include/ll.h:18]: `<type> *const ptr` ensures that function does not change the address supplied by the caller, compiler detects/prevents any bug in the function that could modify contents elsewhere.

Comment (Thalley) [reply subsys/bluetooth/controller/include/ll.h:18]: Right `<type> *const` makes sense to use for `[out]` pointers :)

**Review by Thalley — APPROVED** [review]

**Review by ppryga-nordic — CHANGES_REQUESTED** [review]
Few minor things.

I'm not a fan of abbreviations like `pal` for periodic advertising list. It is hard to figure out what is it until you are in the topic.

File: subsys/bluetooth/controller/hci/hci.c [hunk 22]
```diff
@@ -3341,6 +3344,63 @@ static void le_per_adv_recv_enable(struct net_buf *buf, struct net_buf **evt)
 	ccst = hci_cmd_complete(evt, sizeof(*ccst));
 	ccst->status = status;
 }
+
+#if defined(CONFIG_BT_CTLR_SYNC_PERIODIC_ADV_LIST)
+static void le_add_dev_to_pal(struct net_buf *buf, struct net_buf **evt)
```
[the remark this answers is not in the document]
Comment (Thalley) [reply subsys/bluetooth/controller/hci/hci.c:3349]: The `pal` abbreviation was suggested by me, mainly because the abbreviate makes sense, but also since we updated the whitelist (now called filter accept list) to `fal`.

File: subsys/bluetooth/controller/ll_sw/ull_filter.c [hunk 23]
```diff
@@ -469,9 +518,102 @@ uint8_t ll_priv_mode_set(bt_addr_le_t *id_addr, uint8_t mode)
...
+	 */
+
+	if (addr->type == ADDR_TYPE_ANON) {
+		return 0;
+	}
+
+	return pal_remove(addr, sid);
+}
+
+bool ull_filter_ull_pal_addr_match(const uint8_t addr_type,
```
[the remark this answers is not in the document]
Comment (Thalley) [reply subsys/bluetooth/controller/ll_sw/ull_filter.c:563]: At least the `const uint8_t *` should be used IMO, as that gives useful information to the caller. 

The `* const` has less of an impact IMO, but it does prevent someone from accidentally overwriting the pointer in the function, which does make sense. 

The only argument I have for not using `const uint8_t *const` is that it looks ugly :D

File: subsys/bluetooth/controller/hci/hci.c (hunk 22 above)
[the remark this answers is not in the document]
Comment (ppryga) [reply subsys/bluetooth/controller/hci/hci.c:3349]: Thanks for bringing more context to the topic. In my humble opinion it is a bit confusing. This is another abbreviation that is ok for people with experience in controller programming (are familiar with language/abbreviations used there). It may be intimidating for new people to start contributing to controller because it makes the code reading harder. The learning curve, that is already pretty steep, gets even steeper.

File: subsys/bluetooth/controller/ll_sw/ull_filter.c (hunk 23 above)
[the remark this answers is not in the document]
Comment (ppryga) [reply subsys/bluetooth/controller/ll_sw/ull_filter.c:563]: I have no objections about both (and yes it look ugly ;-)). I just wanted to get consensus whether we start to consider it as recommendation or enforce it.

File: subsys/bluetooth/controller/hci/hci.c (hunk 22 above)
[the remark this answers is not in the document]
Comment (Thalley) [reply subsys/bluetooth/controller/hci/hci.c:3349]: Just checked the 5.3 Core spec to see if the periodic advertising list was ever referenced as "PAL" and it isn't. There is another feature called PAL that was removed in 5.3, otherwise we'd definitely need to change it, but not sure now. 

It is not a strict requirement that if we use an acronym for the filter accept list (FAL) that we should do the same for this list, but to me it makes sense to treat those two somewhat the same. 

@ppryga How do you feel about the FAL acronym then?

[the remark this answers is not in the document]
Comment (ppryga) [reply subsys/bluetooth/controller/hci/hci.c:3349]: @Thalley my feelings about `FAL` are the same as for `PAL`. Is there a `FAL` acronym mentioned in the BT Core spec?

BTW. If there were a PAL used for something else, even though is was removed in 5.3, people may be confused.

[the remark this answers is not in the document]
Comment (Thalley) [reply subsys/bluetooth/controller/hci/hci.c:3349]: From what I can tell, neither FAL or PAL is used in the spec, and both are always written fully.

[the remark this answers is not in the document]
Comment (saleh-unikie) [reply subsys/bluetooth/controller/hci/hci.c:3349]: > 
> 
> Thanks for bringing more context to the topic. In my humble opinion it is a bit confusing. This is another abbreviation that is ok for people with experience in controller programming (are familiar with language/abbreviations used there). It may be intimidating for new people to start contributing to controller because it makes the code reading harder. The learning curve, that is already pretty steep, gets even steeper.

As a person with the least knowledge around Zephyr (compared to others at this discussion), I agree with ppryga. It becomes hard to understand, Also I can see in other files, similar acronyms are not used, especially when "app developers" will use the header file (and is not mainly developed for kernel uses).

File: subsys/bluetooth/controller/hci/hci.c [hunk 24]
```diff
@@ -3341,6 +3344,63 @@ static void le_per_adv_recv_enable(struct net_buf *buf, struct net_buf **evt)
...
+	if (adv_cmds_ext_check(evt)) {
+		return;
+	}
+
+	status = ll_pal_add(&cmd->addr, cmd->sid);
+
+	*evt = cmd_complete_status(status);
+}
+
+static void le_rem_dev_from_pal(struct net_buf *buf, struct net_buf **evt)
```
[the remark this answers is not in the document]
Comment (cvinayak) [reply subsys/bluetooth/controller/hci/hci.c:3363]: I am against use of `PER` as it is abbreviation for Peripheral in the Bluetooth Test Specification.

File: subsys/bluetooth/controller/ll_sw/ull_filter.c [hunk 25]
```diff
@@ -359,7 +399,16 @@ uint8_t ll_rl_remove(bt_addr_le_t *id_addr)
 			fal[j].rl_idx = FILTER_IDX_NONE;
 		}
 
+#if defined(CONFIG_BT_CTLR_SYNC_PERIODIC_ADV_LIST)
```
[the remark this answers is not in the document]
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_filter.c:402]: yes.

File: subsys/bluetooth/controller/hci/hci.c (hunk 24 above)
[the remark this answers is not in the document]
Comment (Thalley) [reply subsys/bluetooth/controller/hci/hci.c:3363]: So if not `pal` and not `per_adv_list`, I guess we are just left with `pa_list` or `periodic_adv_list` (or even `periodic_advertiser_list`). 

We use `per_` for periodic in the Host API though.

File: subsys/bluetooth/controller/hci/hci.c (hunk 22 above)
[the remark this answers is not in the document]
Comment (cvinayak) [reply subsys/bluetooth/controller/hci/hci.c:3349]: I agree, `FAL` and `PAL` are as bad compare to `GATT` or `EATT` or `L2CAP` or `L2CAP ECFC`.

File: subsys/bluetooth/controller/ll_sw/ull_filter.c (hunk 23 above)
[the remark this answers is not in the document]
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_filter.c:563]: This is the prototype for `memcpy`, it isnt ugly to me, but it is something it is.

`void *memcpy(void *restrict dest, const void *restrict src, size_t n);`

File: subsys/bluetooth/controller/hci/hci.c (hunk 24 above)
[the remark this answers is not in the document]
Comment (cvinayak) [reply subsys/bluetooth/controller/hci/hci.c:3363]: Naming is outside the scope of this PR, now that the names have been changed thrice in this PR. 

That said, `per_` as abbreviation in BT API is sad as `per` is a word in itself having different meaning.

I suggest a issue be open to discuss general changes to abbreviations used in the Bluetooth Subsystem, and no change be done without a consensus.

File: subsys/bluetooth/controller/ll_sw/ull_filter.c [hunk 26]
```diff
@@ -152,7 +153,35 @@ static uint8_t prpa_cache_try_resolve(bt_addr_t *rpa);
 static void prpa_cache_resolve(struct k_work *work);
 static void target_resolve(struct k_work *work);
 #endif /* CONFIG_BT_CTLR_SW_DEFERRED_PRIVACY */
+#endif /* CONFIG_BT_CTLR_FILTER_ACCEPT_LIST */
 
+#if defined(CONFIG_BT_CTLR_SYNC_PERIODIC_ADV_LIST)
+#define PAL_ADDR_MATCH(type, addr) \
```
[the remark this answers is not in the document]
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_filter.c:159]: The macro is following a similar `LIST_MATCH` macro used in the filter accept list implementation.

File: subsys/bluetooth/controller/ll_sw/ull_filter.c (hunk 23 above)
[the remark this answers is not in the document]
Comment (ppryga) [reply subsys/bluetooth/controller/ll_sw/ull_filter.c:563]: Concluding, should I consider usage of const pointers wherever possible as recommendation in controller's code or enforce it in my code and in reviews?

File: subsys/bluetooth/controller/ll_sw/ull_filter.c (hunk 26 above)
[the remark this answers is not in the document]
Comment (ppryga) [reply subsys/bluetooth/controller/ll_sw/ull_filter.c:159]: Ok, then let it stay. But for future is seems to me that it would be more beneficial to use functions instead of macros like this one.

File: subsys/bluetooth/controller/ll_sw/ull_filter.c [hunk 27]
```diff
@@ -152,7 +153,35 @@ static uint8_t prpa_cache_try_resolve(bt_addr_t *rpa);
...
+		(pal[i].taken && \
+		 (pal[i].id_addr_type == (type & 0x1)) && \
+		 !memcmp(pal[i].id_addr.val, addr, BDADDR_SIZE))
+
+#define PAL_MATCH(type, addr, sid) \
```
[the remark this answers is not in the document]
Comment (ppryga) [reply subsys/bluetooth/controller/ll_sw/ull_filter.c:164]: Resolving with the same justification as `PAL_ADDR_MATCH`.

File: subsys/bluetooth/controller/hci/hci.c (hunk 24 above)
[the remark this answers is not in the document]
Comment (ppryga) [reply subsys/bluetooth/controller/hci/hci.c:3363]: @cvinayak then please open the issue and after that I'll resolve my comments about naming.

File: subsys/bluetooth/controller/ll_sw/ull_filter.c (hunk 26 above)
[the remark this answers is not in the document]
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_filter.c:159]: Agree, I will make the change as part of the extended advertising filtering implementation changes I am working on currently.

File: subsys/bluetooth/controller/hci/hci.c (hunk 22 above)
[the remark this answers is not in the document]
Comment (ppryga) [reply subsys/bluetooth/controller/hci/hci.c:3349]: The same comment applies here: https://github.com/zephyrproject-rtos/zephyr/pull/38338#discussion_r719292963

File: subsys/bluetooth/controller/ll_sw/ull_filter.c (hunk 23 above)
[the remark this answers is not in the document]
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_filter.c:563]: Internal coding style enforces strict usage of variable qualifiers, but I am open to what is acceptable in the Zephyr code base.

[the remark this answers is not in the document]
Comment (ppryga) [reply subsys/bluetooth/controller/ll_sw/ull_filter.c:563]: Then let we know everything. Qualifiers will be enforced.

File: subsys/bluetooth/controller/hci/hci.c (hunk 24 above)
[the remark this answers is not in the document]
Comment (cvinayak) [reply subsys/bluetooth/controller/hci/hci.c:3363]: https://github.com/zephyrproject-rtos/zephyr/issues/39025

[the remark this answers is not in the document]
Comment (ppryga) [reply subsys/bluetooth/controller/hci/hci.c:3363]: Thank you!

**Review by ppryga-nordic — APPROVED** [review]
